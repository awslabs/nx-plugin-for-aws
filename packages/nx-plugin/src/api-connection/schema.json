{
  "$schema": "https://json-schema.org/schema",
  "$id": "APIConnection",
  "title": "",
  "type": "object",
  "properties": {
    "sourceProject": {
      "type": "string",
      "description": "The source project which will call the API",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Select the source project which will call the API",
      "x-dropdown": "projects"
    },
    "targetProject": {
      "type": "string",
      "description": "The target project containing your API",
      "x-prompt": "Select the target project containing your API",
      "x-dropdown": "projects"
    }
  },
  "required": ["sourceProject", "targetProject"]
}
