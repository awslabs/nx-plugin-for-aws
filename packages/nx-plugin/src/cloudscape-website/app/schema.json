{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    }
  ],
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z][^:]*$"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "default": "packages",
      "x-prompt": "What directory would you like to store your application in?"
    }
  },
  "required": ["name"],
  "examplesFile": "../../../docs/application-examples.md"
}
