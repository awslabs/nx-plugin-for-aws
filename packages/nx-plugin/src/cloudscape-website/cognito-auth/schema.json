{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    },
    {
      "command": "nx g app myapp --classComponent",
      "description": "Use class components instead of functional components"
    }
  ],
  "type": "object",
  "properties": {
    "project": {
      "description": "The root directory of the Cloudscape application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "x-prompt": "the root directory for your Cloudscape application",
      "x-dropdown": "projects"
    },
    "cognitoDomain": {
      "description": "The cognito domain prefix to use",
      "type": "string",
      "x-priority": "important",
      "x-prompt": "the cognito domain prefix to use i.e: https://{{cognitoDomain}}-123456789.auth.ap-southeast-2.amazoncognito.com"
    },
    "allowSignup": {
      "description": "Whether to allow self-signup",
      "type": "boolean",
      "default": false,
      "x-prompt": "Enable self signup?",
      "x-priority": "important"
    }
  },
  "required": ["project", "cognitoDomain"]
}
