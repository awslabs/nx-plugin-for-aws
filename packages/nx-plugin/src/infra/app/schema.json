{
  "$schema": "https://json-schema.org/schema",
  "$id": "Infra",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z][^:]*$"
    },
    "ruleSet": {
      "type": "string",
      "enum": [
        "aws_prototyping",
        "cfn_nag",
        "hipaa",
        "nist_csf",
        "pci_dss_3_2_1",
        "well_architected_reliability",
        "well_architected_security"
      ],
      "x-priority": "important",
      "x-prompt": "What rule set do you want to validate your AWS resources with?",
      "description": "Rule set to validate your AWS resources with.",
      "default": "aws_prototyping"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "default": "packages",
      "x-prompt": "What directory would you like to store your application in?"
    }
  },
  "required": ["name"]
}
