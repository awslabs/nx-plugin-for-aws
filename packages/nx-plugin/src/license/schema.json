{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "License",
  "title": "Add LICENSE files, and configure source code license headers",
  "type": "object",
  "properties": {
    "license": {
      "type": "string",
      "description": "SPDX License Identifier",
      "enum": ["Apache-2.0", "MIT", "ASL"],
      "default": "Apache-2.0",
      "x-priority": "important"
    },
    "copyrightHolder": {
      "type": "string",
      "description": "Copyright holder name",
      "default": "Amazon.com, Inc. or its affiliates",
      "x-priority": "important"
    }
  },
  "required": []
}
