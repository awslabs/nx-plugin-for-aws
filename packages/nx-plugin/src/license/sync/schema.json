{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "$id": "license-headers",
  "title": "Sync generator for adding license headers to source code",
  "type": "object",
  "properties": {},
  "required": []
}
