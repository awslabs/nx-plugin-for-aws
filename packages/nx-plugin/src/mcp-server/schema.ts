/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { z } from 'zod';

export const PACKAGE_MANAGERS = ['pnpm', 'yarn', 'npm', 'bun'] as const;
export const PackageManagerSchema = z.enum(PACKAGE_MANAGERS);
