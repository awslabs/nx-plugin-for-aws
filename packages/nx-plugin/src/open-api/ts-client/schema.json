{
  "$schema": "https://json-schema.org/schema",
  "$id": "OpenApiTsClient",
  "title": "OpenAPI TypeScript Client",
  "description": "Generate TypeScript client from an OpenAPI specification",
  "type": "object",
  "properties": {
    "openApiSpecPath": {
      "type": "string",
      "description": "Path to the OpenAPI specification relative to the monorepo root",
      "x-priority": "important"
    },
    "outputPath": {
      "type": "string",
      "description": "Path to the directory in which to generate the client relative to the monorepo root",
      "x-priority": "important"
    }
  },
  "required": ["openApiSpecPath", "outputPath"]
}
