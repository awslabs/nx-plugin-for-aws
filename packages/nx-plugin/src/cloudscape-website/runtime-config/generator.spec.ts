/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import {
  RUNTIME_CONFIG_GENERATOR_INFO,
  runtimeConfigGenerator,
} from './generator';
import { RuntimeConfigGeneratorSchema } from './schema';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import { expectHasMetricTags } from '../../utils/metrics.spec';

describe('runtime-config generator', () => {
  let tree: Tree;

  const options: RuntimeConfigGeneratorSchema = {
    project: 'test-app',
  };

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
    tree.write(
      'packages/test-app/project.json',
      JSON.stringify({
        name: 'test-app',
        sourceRoot: 'packages/test-app/src',
      }),
    );
  });

  it('should generate runtime config files', async () => {
    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {

          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );
    await runtimeConfigGenerator(tree, options);
    // Check if RuntimeConfig component was generated
    expect(
      tree.exists('packages/test-app/src/components/RuntimeConfig/index.tsx'),
    ).toBeTruthy();
    // Snapshot the generated RuntimeConfig component
    expect(
      tree
        .read('packages/test-app/src/components/RuntimeConfig/index.tsx')
        ?.toString(),
    ).toMatchSnapshot('runtime-config-component.tsx');
  });

  it('should modify main.tsx correctly', async () => {
    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {

          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );
    await runtimeConfigGenerator(tree, options);
    const mainTsxContent = tree
      .read('packages/test-app/src/main.tsx')
      ?.toString();
    expect(mainTsxContent).toMatchSnapshot('modified-main.tsx');
  });

  it('should skip generation if RuntimeConfig already exists', async () => {
    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {

          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );
    // First run to generate files
    await runtimeConfigGenerator(tree, options);
    // Get file content after first run
    const firstRunContent = tree
      .read('packages/test-app/src/components/RuntimeConfig/index.tsx')
      ?.toString();
    // Modify the file to simulate manual changes
    tree.write(
      'packages/test-app/src/components/RuntimeConfig/index.tsx',
      'modified content',
    );
    // Run generator again
    await runtimeConfigGenerator(tree, options);
    // Verify file wasn't overwritten
    expect(
      tree
        .read('packages/test-app/src/components/RuntimeConfig/index.tsx')
        ?.toString(),
    ).toBe('modified content');
    expect(
      tree
        .read('packages/test-app/src/components/RuntimeConfig/index.tsx')
        ?.toString(),
    ).not.toBe(firstRunContent);
  });

  it('should throw error if main.tsx does not exist', async () => {
    await expect(runtimeConfigGenerator(tree, options)).rejects.toThrow(
      'Can only run this generator on a project which contains packages/test-app/src/main.tsx',
    );
  });

  it('should throw error if App is not found', async () => {
    // Set up main.tsx without App
    tree.write(
      'packages/test-app/src/main.tsx',
      `export function App() {
        return <div>Test App</div>;
      }`,
    );
    await expect(runtimeConfigGenerator(tree, options)).rejects.toThrow(
      'Could not locate the App element in main.tsx',
    );
  });

  it('should generate shared constructs', async () => {
    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {

          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );
    await runtimeConfigGenerator(tree, options);
    // Check if shared constructs were generated
    expect(
      tree.exists('packages/common/constructs/src/core/index.ts'),
    ).toBeTruthy();
    expect(
      tree.exists('packages/common/constructs/src/core/runtime-config.ts'),
    ).toBeTruthy();
    // Snapshot the shared constructs files
    expect(
      tree.read('packages/common/constructs/src/core/index.ts')?.toString(),
    ).toMatchSnapshot('common/constructs-index.ts');
    expect(
      tree
        .read('packages/common/constructs/src/core/runtime-config.ts')
        ?.toString(),
    ).toMatchSnapshot('runtime-config.ts');
  });

  it('should add generator metric to app.ts', async () => {
    // Set up test tree with shared constructs
    await sharedConstructsGenerator(tree);

    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {

          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );

    // Call the generator function
    await runtimeConfigGenerator(tree, options);

    // Verify the metric was added to app.ts
    expectHasMetricTags(tree, RUNTIME_CONFIG_GENERATOR_INFO.metric);
  });

  describe('router context integration', () => {
    it('should update RouterProviderContext interface when it exists', async () => {
      // Setup main.tsx with RouterProviderContext interface and createRouter call
      tree.write(
        'packages/test-app/src/main.tsx',
        `
        import { RouterProvider, createRouter } from '@tanstack/react-router';
        import { routeTree } from './routeTree.gen';

        // eslint-disable-next-line @typescript-eslint/no-empty-object-type
        export type RouterProviderContext = {};

        const router = createRouter({
          routeTree,
          context: {}
        });

        const App = () => {
          return <RouterProvider router={router} context={{}} />;
        };

        export function Main() {
          return <App/>;
        }
      `,
      );

      await runtimeConfigGenerator(tree, options);

      const mainTsxContent = tree
        .read('packages/test-app/src/main.tsx')
        .toString();

      // Verify useRuntimeConfig import is added
      expect(mainTsxContent).toContain(
        "import { useRuntimeConfig } from './hooks/useRuntimeConfig'",
      );

      // Verify RouterProviderContext interface is updated with runtimeConfig property
      expect(mainTsxContent).toContain(
        'runtimeConfig?: ReturnType<typeof useRuntimeConfig>',
      );

      // Verify the interface is no longer empty
      expect(mainTsxContent).not.toContain(
        'export type RouterProviderContext = {};',
      );

      // Verify router context is updated
      expect(mainTsxContent).toContain('runtimeConfig: undefined');

      // Verify App component uses hook and passes context to RouterProvider
      expect(mainTsxContent).toContain(
        'const runtimeConfig = useRuntimeConfig();',
      );
      expect(mainTsxContent).toContain(
        '<RouterProvider router={router} context={{ runtimeConfig }} />',
      );
    });

    it('should update router context when createRouter exists', async () => {
      // Setup main.tsx with createRouter call but no RouterProviderContext
      tree.write(
        'packages/test-app/src/main.tsx',
        `
        import { RouterProvider, createRouter } from '@tanstack/react-router';
        import { routeTree } from './routeTree.gen';

        export type RouterProviderContext = {};

        const router = createRouter({
          routeTree,
          context: {}
        });

        const App = () => {
          return <RouterProvider router={router} context={{}} />;
        };

        export function Main() {
          return <App/>;
        }
      `,
      );

      await runtimeConfigGenerator(tree, options);

      const mainTsxContent = tree
        .read('packages/test-app/src/main.tsx')
        .toString();

      // Verify router context includes runtimeConfig
      expect(mainTsxContent).toContain('runtimeConfig: undefined');
    });

    it('should update App component when it exists', async () => {
      // Setup main.tsx with App component
      tree.write(
        'packages/test-app/src/main.tsx',
        `
        import { RouterProvider, createRouter } from '@tanstack/react-router';
        import { routeTree } from './routeTree.gen';

        export type RouterProviderContext = {};

        const router = createRouter({
          routeTree,
          context: {}
        });

        const App = () => {
          return <RouterProvider router={router} context={{}} />;
        };

        export function Main() {
          return <App/>;
        }
      `,
      );

      await runtimeConfigGenerator(tree, options);

      const mainTsxContent = tree
        .read('packages/test-app/src/main.tsx')
        .toString();

      // Verify App component uses useRuntimeConfig hook
      expect(mainTsxContent).toContain(
        'const runtimeConfig = useRuntimeConfig();',
      );

      // Verify RouterProvider receives the context value
      expect(mainTsxContent).toContain(
        '<RouterProvider router={router} context={{ runtimeConfig }} />',
      );

      // Verify the App component is now a proper function with a block body
      expect(mainTsxContent).toMatch(/const App = \(\) => \{[\s\S]*\}/);
    });

    it('should handle cases where router context elements do not exist', async () => {
      // Setup main.tsx without router context elements
      tree.write(
        'packages/test-app/src/main.tsx',
        `
        export function Main() {
          const App = () => <div>Simple App</div>;
          return <App/>;
        }
      `,
      );

      // Should not throw error when router elements don't exist
      await expect(
        runtimeConfigGenerator(tree, options),
      ).resolves.not.toThrow();

      const mainTsxContent = tree
        .read('packages/test-app/src/main.tsx')
        .toString();

      expect(mainTsxContent).not.toContain('useRuntimeConfig()');
    });
  });
});
