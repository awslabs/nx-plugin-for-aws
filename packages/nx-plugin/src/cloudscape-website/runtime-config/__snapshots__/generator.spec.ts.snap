// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runtime-config generator > should generate runtime config files > runtime-config-component.tsx 1`] = `
"import { Alert, Spinner } from '@cloudscape-design/components';
import React, {
  createContext,
  PropsWithChildren,
  useEffect,
  useState,
} from 'react';
import type { IRuntimeConfig } from ':proj/common-types';

/**
 * Context for storing the runtimeConfig.
 */
export const RuntimeConfigContext = createContext<IRuntimeConfig | undefined>(
  undefined,
);

/**
 * Sets up the runtimeConfig.
 *
 * This assumes a runtime-config.json file is present at '/'.
 */
const RuntimeConfigProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [runtimeConfig, setRuntimeConfig] = useState<
    IRuntimeConfig | undefined
  >();
  const [error, setError] = useState<string | undefined>();

  useEffect(() => {
    fetch('/runtime-config.json')
      .then((response) => {
        return response.json();
      })
      .then((_runtimeConfig) => {
        setRuntimeConfig(_runtimeConfig as IRuntimeConfig);
      })
      .catch(() => {
        setError('No runtime-config.json detected');
      });
  }, [setRuntimeConfig]);

  return error ? (
    <Alert type="error" header="Validation error">
      {error}
    </Alert>
  ) : runtimeConfig ? (
    <RuntimeConfigContext.Provider value={runtimeConfig}>
      {children}
    </RuntimeConfigContext.Provider>
  ) : (
    <Spinner />
  );
};

export default RuntimeConfigProvider;
"
`;

exports[`runtime-config generator > should generate shared constructs > common/constructs-index.ts 1`] = `
"export * from './runtime-config.js';
"
`;

exports[`runtime-config generator > should generate shared constructs > runtime-config.ts 1`] = `
"import type { IRuntimeConfig } from ':proj/common-types';
import { Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

const RuntimeConfigKey = '__RuntimeConfig__';

export class RuntimeConfig extends Construct {
  private readonly _runtimeConfig: Partial<IRuntimeConfig> = {};

  static ensure(scope: Construct): RuntimeConfig {
    const stack = Stack.of(scope);
    return (
      RuntimeConfig.of(scope) ?? new RuntimeConfig(stack, RuntimeConfigKey)
    );
  }

  static of(scope: Construct): RuntimeConfig | undefined {
    const stack = Stack.of(scope);
    return stack.node.tryFindChild(RuntimeConfigKey) as
      | RuntimeConfig
      | undefined;
  }

  constructor(scope: Construct, id: string) {
    super(scope, id);
  }

  get config(): Partial<IRuntimeConfig> {
    return this._runtimeConfig;
  }
}
"
`;

exports[`runtime-config generator > should modify main.tsx correctly > modified-main.tsx 1`] = `
"import RuntimeConfigProvider from './components/RuntimeConfig';
import { RouterProvider } from '@tanstack/react-router';

export function App() {
  return (
    <RuntimeConfigProvider>
      <RouterProvider router={router} />
    </RuntimeConfigProvider>
  );
}
"
`;
